src/SGFParser.cpp

```diff
@@ -38,6 +38,81 @@
 #include "SGFTree.h"
 #include "Utils.h"
 
+std::vector<SGFParser::ELF_Data> SGFParser::chop_elf(std::string filename,
+                                                     size_t stopat) {
+    std::ifstream ins(filename.c_str(), std::ifstream::binary | std::ifstream::in);
+
+    if (ins.fail()) {
+        throw std::runtime_error("Error opening file");
+    }
+
+    std::vector<ELF_Data> result;
+    std::string gamebuff;
+
+    auto constexpr MAX = std::numeric_limits<std::streamsize>::max();
+    std::string str;
+    char c;
+    int num_move;
+    int total_policy;
+
+    while (result.size() <= stopat) {
+        while (true) {
+            ins.ignore(MAX, '\"');
+            if ((ins.rdstate() & std::ifstream::eofbit) != 0) return result;
+            std::getline(ins, str, ':');
+            if (str == "content\"") break;
+        }
+        result.emplace_back(ELF_Data());
+        auto& data = result.back();
+        ins >> c; // eat "
+        str.clear();
+        std::getline(ins, str, '\"');
+        data.sgf = str;
+        ins.ignore(MAX, ':'); // eat ,"num_move":
+        str.clear();
+        std::getline(ins, str, ',');
+        num_move = std::stoi(str);
+        ins.ignore(MAX, '['); // eat "policies":[
+        data.policies.clear();
+        data.policies.reserve(num_move + 1);
+        while (true) {
+            ins >> c; // eat [ (assume there's at least one policy record)
+            str.clear();
+            std::getline(ins, str, ',');
+            data.policies.emplace_back(std::array<float, POTENTIAL_MOVES>());
+            auto& probs = data.policies.back();
+            probs[NUM_INTERSECTIONS] = total_policy = std::stoi(str);
+            for (auto i = 0; i < BOARD_SIZE; i++) ins.ignore(MAX, ',');
+            for (int i = BOARD_SIZE - 1; i >= 0; i--) {
```

Comment:
Weird mix of auto and int here.

Reply:
With auto `i` will be assigned an `unsigned` type, and will always be `>=0` leading to infinite loop. I could change it to `auto i = (int)BOARD_SIZE - 1;` if you wish...